src/lib/tuning/tuner.hpp

```diff
@@ -0,0 +1,154 @@
+#pragma once
+
+#include <chrono>
+#include <limits>
+#include <memory>
+#include <string>
+
+#include "scheduler/abstract_task.hpp"
+#include "tuning/abstract_evaluator.hpp"
+#include "tuning/abstract_selector.hpp"
+
+namespace opossum {
+
+/**
+ * A Tuner encapsulates the process of analyzing the current system state and
+ * performing specific modifications to optimize the systems performance.
+ *
+ * It uses AbstractEvaluators to generate TuningChoices which represent possible
+ * modifications together with their expected performance impact and costs.
+ * TuningChoices are transformed into a concrete operation sequence
+ * by an AbstractSelector, that also considers a cost budget that may not be
+ * exceeded at any point in the sequence.
+ *
+ * While executing the operation sequence, the Tuner considers a time budget
+ * and stops the execution once it exceeds that budget.
+ *
+ * Both the time and cost budget (the latter being enforced by the AbstractSelector)
+ * are initially infinite (i.e. disabled).
+ */
+class Tuner {
+ public:
+  using Runtime = std::chrono::duration<float, std::chrono::seconds::period>;
+  using RuntimeClock = std::chrono::high_resolution_clock;
+
+  enum class Status { Unknown, Running, Completed, EvaluationTimeout, SelectionTimeout, ExecutionTimeout, Timeout };
+
+  static constexpr float NoBudget = std::numeric_limits<float>::infinity();
+
+  Tuner();
+
+  /**
+   * Accessors for the AbstractEvaluators used for the tuning process.
+   *
+   * The Tuner takes ownership of the supplied evaluator. The same evaluator
+   * instance may never be used in more than one Tuner.
+   */
+  void add_evaluator(std::unique_ptr<AbstractEvaluator>&& evaluator);
+  void remove_evaluator(std::size_t index);
+  const std::vector<std::unique_ptr<AbstractEvaluator>>& evaluators() const;
+
+  /**
+   * Accessors for the AbstractSelector used for the tuning process.
+   *
+   * The Tuner takes ownership of the supplied selector. The same selector
+   * instance may never be used in more than one Tuner.
+   */
+  void set_selector(std::unique_ptr<AbstractSelector>&& selector);
+  const std::unique_ptr<AbstractSelector>& selector() const;
+
+  /**
+   * Configures the time budgets for subsequent schedule() invocations.
+   *
+   * evaluate_budget: time budget for the evaluation phase
+   * select_budget: time budget for the selection phase
+   * execute_budget: time budget for the evaluation phase
+   * budget: overall budget for all three phases
```

Comment:
Why do we need this? Isn‘t this `evaluate + select + execute`? I can imagine situation where it makes sense to keep the overall budget lower than the sum, but that should be explained here

Reply:
the three phase budgets are orthogonal to the overall budget, to support limiting the entire tuning process, any specific phase or any combination of the former. As we can not predict what particular time budget might be relevant for a future use case, this _is_ needed. Especially as it is easy to disable any one of those budgets.

We can however stress the point in the comment, that all four budgets are checked separately and that aborting one phase happens as the disjunction of its particular phase timeout and the overall timeout.